staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/conversion.go

```diff
@@ -41,3 +41,72 @@ func ConvertToGVK(obj runtime.Object, gvk schema.GroupVersionKind, o admission.O
 	out.GetObjectKind().SetGroupVersionKind(gvk)
 	return out, nil
 }
+
+// NewVersionedAttributes returns versioned attributes with the old and new object (if non-nil) converted to the requested kind
+func NewVersionedAttributes(attr admission.Attributes, gvk schema.GroupVersionKind, o admission.ObjectInterfaces) (*VersionedAttributes, error) {
+	// convert the old and new objects to the requested version
+	versionedAttr := &VersionedAttributes{
+		Attributes:    attr,
+		VersionedKind: gvk,
+	}
+	if oldObj := attr.GetOldObject(); oldObj != nil {
+		out, err := ConvertToGVK(oldObj, gvk, o)
+		if err != nil {
+			return nil, err
+		}
+		versionedAttr.VersionedOldObject = out
+	}
+	if obj := attr.GetObject(); obj != nil {
+		out, err := ConvertToGVK(obj, gvk, o)
+		if err != nil {
+			return nil, err
+		}
+		versionedAttr.VersionedObject = out
+	}
+	return versionedAttr, nil
+}
+
+// ConvertVersionedAttributes converts VersionedObject and VersionedOldObject to the specified kind, if needed.
+// If attr.VersionedKind already matches the requested kind, no conversion is performed.
+// If conversion is required:
+// * attr.VersionedObject is used as the source for the new object if Dirty=true (and is round-tripped through attr.Attributes.Object, clearing Dirty in the process)
+// * attr.Attributes.Object is used as the source for the new object if Dirty=false
+// * attr.Attributes.OldObject is used as the source for the old object
+func ConvertVersionedAttributes(attr *VersionedAttributes, gvk schema.GroupVersionKind, o admission.ObjectInterfaces) error {
+	// we already have the desired kind, we're done
+	if attr.VersionedKind == gvk {
+		return nil
+	}
+
+	// convert the original old object to the desired GVK
+	if oldObj := attr.Attributes.GetOldObject(); oldObj != nil {
+		out, err := ConvertToGVK(oldObj, gvk, o)
+		if err != nil {
+			return err
+		}
+		attr.VersionedOldObject = out
+	}
+
+	if attr.VersionedObject != nil {
+		// convert the existing versioned object to internal
+		if attr.Dirty {
+			err := o.GetObjectConvertor().Convert(attr.VersionedObject, attr.Attributes.GetObject(), nil)
```

Comment:
this case feels asymmetric: if `attr.VersionedKind == gvk` we don't write back to `GetObject()´. But if we convert, we do. Why don't we keep converting from versioned object to versioned object, keeping attr dirty?

Reply:
> this case feels asymmetric: if attr.VersionedKind == gvk we don't write back to `GetObject()´. But if we convert, we do. Why don't we keep converting from versioned object to versioned object, keeping attr dirty?

The goal of this method is to provide a VersionedObject in the requested version as cheaply as possible. If we already have the requested version, we no-op.

If we don't have the requested verison, we *have* to convert via `attr.GetObject()` because that's our internal version (we can't convert directly from extension/v1beta1 Deployment to apps/v1 Deployment). As long as we're updating attr.GetObject() to match VersionedObject anyway, we might as well clear the dirty flag to avoid needing to redo the same conversion at the end of the mutating Dispatch() method

I'll update the docs on the fields and this method to try to clarify that

Reply:
right. Am too much in the CRD world without internal versions. Nvmd.